### PR TITLE
Feat/schema management v2

### PR DESCRIPTION
**Pull request by loicgreffier**

### CODE REVIEW

**Review by twobeeb — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/michelin/ns4kafka/controllers/SchemaControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+        Assertions.assertEquals(1L, exception.getValidationErrors().size());
+        Assertions.assertEquals("Invalid prefix prefix.subject : namespace not owner of this subject", exception.getValidationErrors().get(0));
+        verify(this.schemaService, never()).updateSubjectCompatibility(namespace, "prefix.subject", Schema.Compatibility.BACKWARD);
+    }
+
+    /**
+     * Test the compatibility update in dry mode
+     */
+    @Test
+    void compatibilityUpdateDryRun() {
```
[the remark this answers is not in the document]
Comment (loicgreffier) [reply api/src/test/java/com/michelin/ns4kafka/controllers/SchemaControllerTest.java:303]: Fixed on the compatibilityUpdate() test

File: api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+     * Update the compatibility of a subject
+     *
+     * @param namespace The namespace
+     * @param subject The subject to update
+     * @param compatibility The compatibility to apply
+     * @param dryrun Dry run mode
+     * @return The last updated schema
+     */
+    @Post("/{subject}/compatibility")
+    public HttpResponse<Optional<Schema>> compatibility(String namespace, @PathVariable String subject, @Valid @Body Map<String, Schema.Compatibility> compatibility,
```
Comment (twobeeb) [inline api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java:146]: Simple body maps to JSON directly
```suggestion
    public HttpResponse<Optional<Schema>> compatibility(String namespace, @PathVariable String subject, @Body Schema.Compatibility compatibility,
```

Comment (loicgreffier) [reply api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java:147]: Mapping a simple String to Schema.Compatibility directly seems not working with JSON format

**Comment by twobeeb** [comment]
Need to verify that objects are always returned augmented with
````
getMetadata().getName()
getMetadata().getNamespace()
getMetadata().getCluster()
````
before any calls to `sendEventLog()`

File: api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                    " : namespace not owner of this subject"), AccessControlEntry.ResourceType.SCHEMA.toString(), subject);
+        }
+
+        Optional<Schema> existingSchemaOptional = this.schemaService
+                .getLatestSubject(ns, subject);
+
+        if (existingSchemaOptional.isEmpty()) {
+            return HttpResponse.notFound();
+        }
+
```
Comment (twobeeb) [inline api/src/main/java/com/michelin/ns4kafka/controllers/SchemaController.java:177]: Something along the line of:
```java
sendEventLog("SchemaCompatibilityState",
                existingSchemaOptional.get().getMetadata(),
                ApplyStatus.changed,
                existingSchemaOptional.get().getSpec().getCompatibility,
                compatibility);
```

File: cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java [hunk 4]
```diff
@@ -71,14 +76,25 @@ public Resource getSingleResourceWithType(ApiResource apiResource, String namesp
...
                 return nonNamespacedClient.apply(loginService.getAuthorization(), apiResource.getPath(), resource, dryRun);
             }
         } catch (HttpClientResponseException e) {
             formatService.displayError(e, apiResource.getKind(), resource.getMetadata().getName());
+        } catch (NoSuchFileException e) {
+            System.out.printf("Failed : %s/%s %s%n", apiResource.getKind(), resource.getMetadata().getName(),
+                    "Cannot find schema. Schema path must be relative to the CLI, not to the resource file");
+        } catch (IOException e) {
+            throw new RuntimeException(e);
         }
```
Comment (twobeeb) [inline cli/src/main/java/com/michelin/ns4kafka/cli/services/ResourceService.java:96]: Throw an exception with your custom message in fileService.loadSchemaFileContent and use ``formatService.displayError`` instead:

```suggestion
        } catch (HttpClientResponseException | YourCustomFileLoadingException e) {
            formatService.displayError(e, apiResource.getKind(), resource.getMetadata().getName());
        }
```

**Review by twobeeb — CHANGES_REQUESTED** [review]

**Pull request closed without merging**